### PR TITLE
Keep ifcfgs provided via linuxrc when keep_install_network is set in AY profile

### PR DESCRIPTION
**Pull request by mchf**
https://bugzilla.suse.com/show_bug.cgi?id=874259

### CODE REVIEW

File: src/clients/lan_auto.ycp [hunk 1]
```diff
@@ -62,6 +62,65 @@ map UpcaseCondSet (map ret, map m, string key) {
...
+    y2milestone("MergeDevices: start (in_devs1: %1, in_devs2: %2)", in_devs1, in_devs2);
+
+    if(in_devs1 == nil || in_devs1 == {})
+    {
+        merged_devices = in_devs2 == nil ? $[] : in_devs2;
+        return merged_devices;
+    }
+    if(in_devs2 == nil || in_devs2 == {})
+    {
+        merged_devices = in_devs1 == nil ? $[] : in_devs1;
```
Comment (jreidinger) [inline src/clients/lan_auto.ycp]: cannot happen so simple `merged_devices = in_devs1;` is enough

File: src/clients/lan_auto.ycp [hunk 2]
```diff
@@ -62,6 +62,65 @@ map UpcaseCondSet (map ret, map m, string key) {
...
+    merged_devices = in_devs1;
+
+    y2milestone("MergeDevices: going to do real merge");
+
+    foreach(string type, any devices, in_devs2,
+    {
+        if(haskey(in_devs1, type))
+        {
+            map merged = union((map) in_devs1[type]:{}, (map) in_devs2[type]:{});
+            merged_devices = add(merged_devices, type, merged);
```
Comment (jreidinger) [inline src/clients/lan_auto.ycp]: why not simple `merged_devices[type] = merged`???

File: src/clients/lan_auto.ycp [hunk 3]
```diff
@@ -312,10 +369,10 @@ else if(func == "Import") {
         if (! haskey(param, "dns")) {
             param["dns"] = $[];
         }
-        foreach( string key, any value, dns, {
-            
+        foreach( string key, any value, dns,
+        {
             if( ! haskey( param["dns"]:$[], key )
-                && key != "write_hostname" ) { // [1] ^
```
Comment (mvidner) [inline src/clients/lan_auto.ycp]: If this "[1]" is deleted then so should be the one 10 lines earlier.

**Comment by jreidinger** [comment]
in general nothing critical that blocks merging, just few notes

File: src/clients/lan_auto.ycp (hunk 2 above)
Comment (mchf) [inline src/clients/lan_auto.ycp]: honestly I don't know. May be I'm so used to ```Builtins.add``` from ruby branch that I've used that even here ;-)

Will fix it.

File: src/clients/lan_auto.ycp (hunk 1 above)
Comment (mchf) [inline src/clients/lan_auto.ycp]: copy-paste programming ;-)

File: src/clients/lan_auto.ycp [hunk 4]
```diff
@@ -152,8 +208,9 @@ define map FromAY(map input)
         });
     }
 
-
-    input["devices"] = devices;
+    // merge devices definitions obtained from inst-sys (input["devices"])
+    // and those which were read from AY profile. bnc#874259
+    input["devices"] = MergeDevices((map<string, any>) input["devices"]:{}, devices);
```
Comment (mvidner) [inline src/clients/lan_auto.ycp]: `$[]`, not `{}`.

Comment (mchf) [inline src/clients/lan_auto.ycp]: thanks

File: src/clients/lan_auto.ycp [hunk 5]
```diff
@@ -62,6 +62,65 @@ map UpcaseCondSet (map ret, map m, string key) {
...
+ *
+ * @return merged device map in NetworkInterfaces format or empty map
+ */
+define map MergeDevices(map<string, any> in_devs1, map<string, any> in_devs2)
+{
+    map merged_devices = $[];
+
+    y2milestone("MergeDevices: start (in_devs1: %1, in_devs2: %2)", in_devs1, in_devs2);
+
+    if(in_devs1 == nil || in_devs1 == {})
```
Comment (mvidner) [inline src/clients/lan_auto.ycp]: $[]

File: src/clients/lan_auto.ycp [hunk 6]
```diff
@@ -62,6 +62,65 @@ map UpcaseCondSet (map ret, map m, string key) {
...
+
+    merged_devices = in_devs1;
+
+    y2milestone("MergeDevices: going to do real merge");
+
+    foreach(string type, any devices, in_devs2,
+    {
+        if(haskey(in_devs1, type))
+        {
+            map merged = union((map) in_devs1[type]:{}, (map) in_devs2[type]:{});
```
Comment (mvidner) [inline src/clients/lan_auto.ycp]: `$[]`, and here the need to cast to a generic `map` should really tip us off.

File: src/clients/lan_auto.ycp [hunk 7]
```diff
@@ -62,6 +62,65 @@ map UpcaseCondSet (map ret, map m, string key) {
...
+        return merged_devices;
+    }
+    if(in_devs2 == nil || in_devs2 == {})
+    {
+        merged_devices = in_devs1 == nil ? $[] : in_devs1;
+        return merged_devices;
+    }
+
+    merged_devices = in_devs1;
+
```
Comment (mvidner) [inline src/clients/lan_auto.ycp]: All of the logic above(!) can be replaced with
```c
if (in_devs1 == nil)
{
  in_devs1 = $[];
}
if (in_devs2 == nil)
{
  in_devs2 = $[];
}
```

Comment (mchf) [inline src/clients/lan_auto.ycp]: matter of taste ... I prefer early return

File: src/clients/lan_auto.ycp (hunk 3 above)
Comment (mchf) [inline src/clients/lan_auto.ycp]: done

**Comment by mvidner** [comment]
LGTM.

**Comment by mchf** [comment]
Thanks for reviews.

**Pull request merged**